src/QSymbolicsBase/predefined.jl

```diff
@@ -286,7 +286,109 @@ function Base.show(io::IO, x::SDagger)
     print(io,x.obj)
     print(io,"†")
 end
-symbollabel(x::SDagger) = symbollabel(x.obj)
+
+"""Trace of an operator
+
+```jldoctest
+julia> @op A; @op B;
+
+julia> tr(A)
+tr(A)
+
+julia> tr(commutator(A, B))
+0
+
+julia> @bra b; @ket k;
+
+julia> tr(k*b)
+⟨b||k⟩
+```
+"""
+@withmetadata struct STrace <: Symbolic{Complex}
+    op::Symbolic{AbstractOperator}
+end
+isexpr(::STrace) = true
+iscall(::STrace) = true
+arguments(x::STrace) = [x.op]
+sorted_arguments(x::STrace) = arguments(x)
+operation(x::STrace) = tr
+head(x::STrace) = :tr
+children(x::STrace) = [:tr, x.op]
+Base.show(io::IO, x::STrace) = print(io, "tr($(x.op))")
+tr(x::Symbolic{AbstractOperator}) = STrace(x)
+tr(x::SScaled{AbstractOperator}) = x.coeff*tr(x.obj)
+tr(x::SAdd{AbstractOperator}) = (+)((tr(i) for i in arguments(x))...)
+tr(x::SOuterKetBra) = x.bra*x.ket
+tr(x::SCommutator) = 0
+tr(x::STensorOperator) = (*)((tr(i) for i in arguments(x))...)
+Base.hash(x::STrace, h::UInt) = hash((head(x), arguments(x)), h)
+Base.isequal(x::STrace, y::STrace) = isequal(x.op, y.op)
+
+"""Partial trace over system i of a composite quantum system
+
+```jldoctest
+julia> @op A; @op B;
+
+julia> ptrace(A, 1)
+tr1(A)
+
+julia> ptrace(A⊗B, 1)
+(tr(A))B
+
+julia> @ket k; @bra b;
+
+julia> pure_state = A ⊗ (k*b)
+(A⊗|k⟩⟨b|)
+
+julia> ptrace(pure_state, 1)
+(tr(A))|k⟩⟨b|
+
+julia> ptrace(pure_state, 2)
+(⟨b||k⟩)A
+
+julia> mixed_state = (A⊗(k*b)) + ((k*b)⊗B)
+((A⊗|k⟩⟨b|)+(|k⟩⟨b|⊗B))
+
+julia> ptrace(mixed_state, 1)
+((0 + ⟨b||k⟩)B+(tr(A))|k⟩⟨b|)
+
+julia> ptrace(mixed_state, 2)
+((0 + ⟨b||k⟩)A+(tr(B))|k⟩⟨b|)
+```
+"""
+@withmetadata struct SPartialTrace <: Symbolic{Complex}
+    obj
+    sys::Int
+end
+isexpr(::SPartialTrace) = true
+iscall(::SPartialTrace) = true
+arguments(x::SPartialTrace) = [x.obj, x.sys]
+operation(x::SPartialTrace) = ptrace
+head(x::SPartialTrace) = :ptrace
+children(x::SPartialTrace) = [:ptrace, x.obj, x.sys]
+Base.show(io::IO, x::SPartialTrace) = print(io, "tr$(x.sys)($(x.obj))")
+ptrace(x::Symbolic{AbstractOperator}, s) = SPartialTrace(x, s)
+function ptrace(x::STensorOperator, s)
+    terms = arguments(x)
+    sys_op = terms[s]
+    new_terms = deleteat!(copy(terms), s)
+    isone(length(new_terms)) ? tr(sys_op)*first(new_terms) : tr(sys_op)*STensorOperator(new_terms)
+end
```

Comment:
This is actually a bit worse than this. You can have `@op A SpinBasis(1//2)^2`, i.e. the symbolic operator is already over a composite basis. The number of subsystem would be the number of subsystems in `basis`, which might be different from the number of terms in the tensor product.

Also, what if flattening has not happened. E.g `ptrace(A⊗(B⊗C+D⊗E), 2)`?

Reply:
Ah, these are both fair comments. Will make changes.

Reply:
Just found another small bug:
```
julia> @op A; @op B; @op C;

julia> tp = A⊗B⊗C
((A⊗B)⊗C)

julia> QuantumSymbolics.arguments(tp)
2-element Vector{Any}:
 (A⊗B)
 C
```
We would want `tp` to show `(A⊗B⊗C)` and `QuantumSymbolics.arguments(tp)` to be a 3-element vector of the symbolic operators.

Weird. I wonder why this doesn't happen for multiplication of operators:
```
julia> mp = A*B*C
ABC

julia> QuantumSymbolics.arguments(mp)
3-element Vector{Any}:
 A
 B
 C
```

Reply:
> Also, what if flattening has not happened. E.g `ptrace(A⊗(B⊗C+D⊗E), 2)`?
 
A question somewhat related to this: should we have a function in the package that expands expressions? For instance, if a user for some reason wanted to convert `A⊗(B⊗C+D⊗E)` into `A⊗B⊗C+A⊗D⊗E`, shouldn't they have a way to do that? Maybe we could call it `qexpand`, using the `q` prefix the same we do for `qsimplify`.

Reply:
yup, let's have a `qexpand` for now. Maybe at some point we will just hook into the standard `expand` and at that point it can be just an alias without making a breaking release

concerning the difference between mul and tensor, I think it is because `prefactorscalings` were created for tensor, and it has been kinda hackishly applied to mul. We probably need to clean `prefactorscalings` to work on both. I will make a separate issue for that.

Reply:
I created but then deleted the issue because I had misunderstood something when I was writing it. Could you remind me why you added `scalars=false` options to `prefactorscalings`? I do not think I understand it. I see it has something to do with `SCommutator` but I do not get it.

Reply:
To finally actually answer your question about why * and ⊗ behave differently in terms of flattening, it is because of the Julia parser. Julia parses repeated * as a single multi-argument expression, while it parses ⊗ as a nested expression. The author of the parser has expressed uncertainty whether that was a good decision. One reason it can in principle be useful, is because it permits optimizing the order of matrix multiplications. But maybe it should not have been special cased, rather done for all infix operators.

```
julia> :(a*b*c) |> dump
Expr
  head: Symbol call
  args: Array{Any}((4,))
    1: Symbol *
    2: Symbol a
    3: Symbol b
    4: Symbol c

julia> :(a⊗b⊗c) |> dump
Expr
  head: Symbol call
  args: Array{Any}((3,))
    1: Symbol ⊗
    2: Expr
      head: Symbol call
      args: Array{Any}((3,))
        1: Symbol ⊗
        2: Symbol a
        3: Symbol b
    3: Symbol c

```

Reply:
besides simplifying and hopefully removing the `scalar=false` from prefactorscaling, maybe we should add a simple `flattenop` function that does

`flattenop(*, [A, B*C, D, E⊗F, G+H])` = [A, B, C, D, E⊗F, G+H]
`flattenop(+, [A, B*C, D, E⊗F, G+H])` = [A, B*C, D, E⊗F, G, H]
`flattenop(⊗, [A, B*C, D, E⊗F, G+H])` = [A, B*C, D, E, F, G+H]

and pretty much always use them together

Reply:
I rewrote `scalar=false` options to `prefactorscalings` so that I could pull only a scalar out from a subexpression rather than, say an entire operator. This was important with expressions such as commutators, where if I had the commutator of a scalar operator, I could factor out the scalar value. For instance, `commutator(2*A, B)` would give `2*commutator(A, B)`. In retrospect, this is a messy fix, so that's my bad. I will dedicate time to cleaning up the flattening procedures.

Reply:
Could you give an example where the commutator would misbehave if the `scalar` keyword did not exist?

Reply:
If the `scalar` keyword did not exist, then for example, `commutator(A*B, C)` would give `A*[B,C]`, which is different than the correct property `[A,C]*B + A*[B,C]`.

Reply:
what if in `commutator` we do not do `prefactorscalings([op1, op2])` but rather do:

```
c1, t1 = prefactorscalings(op1)
head1 = length(t1)==1 ? identity : operation(op1)
c2, t2...
c1*c2 * new(head1(c1...), head2(c2...))
```

Reply:
Would it be good to create another file that handles utilities like `flattenop` and `prefactorscalings`, perhaps `utils.jl`? As it doesn't make much sense to have them in `rules.jl` where we are defining simplification rules.

Reply:
Wait, no that is overly complicated. Why not just have `scalar=true` always. Is there a situation that benefits from `scalar=false`?

Reply:
> what if in `commutator` we do not do `prefactorscalings([op1, op2])` but rather do:
> 
> ```
> c1, t1 = prefactorscalings(op1)
> head1 = length(t1)==1 ? identity : operation(op1)
> c2, t2...
> c1*c2 * new(head1(c1...), head2(c2...))
> ```
Yes that sounds good.

Reply:
yes about a new file

Reply:
It's my understanding that `prefactorscalings` was originally created for tensor product properties. For instance, `(A|k1⟩)⊗|k2⟩` would give `A*(k1⟩⊗|k2⟩)`.

Reply:
Regardless, I will clean up the scaling and flattening functions.

Reply:
`prefactorscaling` was always meant only for scalars. The operator identity you are giving is wrong because A is defined to act on the Hilbert space of k2, not on the composite hilbert space of k1k2.